include/oc_push.h

```diff
@@ -0,0 +1,203 @@
+#define OC_PUSH_H
+
+#include <stdio.h>
+
+#include "oc_config.h"
+#include "oc_helpers.h"
+#include "oc_rep.h"
+#include "oc_ri.h"
+#include "oc_endpoint.h"
+#include "port/oc_log.h"
+#include "util/oc_memb.h"
+#include "util/oc_process.h"
+
+
+#if defined(OC_PUSHDEBUG) || defined(OC_DEBUG)
+#ifndef OC_DEBUG
+#define oc_print(...) printf(__VA_ARGS__)
+#define oc_log(level, ...)                                                     \
+		do {                                                                         \
+			oc_print("%s: %s <%s:%d>: ", level, __FILE__, __func__, __LINE__);            \
+			oc_print(__VA_ARGS__);                                                        \
+			oc_print("\n");                                                               \
+		} while (0)
+#define p_dbg(...) oc_log("DEBUG", __VA_ARGS__)
+#define p_wrn(...) oc_log("WARNING", __VA_ARGS__)
+#define p_err(...) oc_log("ERROR", __VA_ARGS__)
+#else
+#define p_dbg(...) OC_LOG("DEBUG", __VA_ARGS__)
+#define p_wrn(...) OC_LOG("WARNING", __VA_ARGS__)
+#define p_err(...) OC_LOG("ERROR", __VA_ARGS__)
+#endif
+#else
+#define p_dbg(...)
+#define p_wrn(...)
+#define p_err(...)
+#endif
+
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+
+#define PUSHCONF_RSC_PATH "/pushconfig"
+#define PUSHCONF_RSC_TYPE "oic.r.pushconfiguration"
+#define PUSHCONF_RSC_NAME "Push Configuration"
+
+#define PUSHRECVS_RSC_PATH "/pushreceivers"
+#define PUSHRECVS_RSC_TYPE "oic.r.pushreceiver"
+#define PUSHRECVS_RSC_NAME "Push Receiver Configuration"
+
+
+typedef enum {
+	OC_PP_WFP,
+	OC_PP_WFU,
+	OC_PP_WFR,
+	OC_PP_WFUM,
+	OC_PP_WFRM,
+	OC_PP_ERR,
+	OC_PP_TOUT
+} oc_pp_state_t;
+
+/*
+ * Resource creation and request handlers for ["oic.r.notificationselector", "oic.r.pushproxy"] Resource
+ */
+typedef struct oc_ns
+{
+	struct oc_ns *next;
+	oc_resource_t *resource;
+	/* notificaiton selector */
+	oc_string_t phref;		/* optional */
+	oc_string_array_t prt;	/* optional */
+	oc_string_array_t pif;	/* optional */
+	/* push proxy */
+	oc_string_t pushtarget_di; /* device id of target (e.g. ocf://17087f8c-13e3-4849-4258-65af2a47df63 */
+	oc_endpoint_t pushtarget_ep;	/* full URI (e.g. coaps://[fe80::b1d6]:1122/myLightSwitch), oc_endpoint_t type */
+	oc_string_t targetpath; /* path in target server (e.g. /myLightSwitch) */
+	oc_string_t pushqif;
+	oc_string_array_t sourcert;
+	oc_string_t state;
+	void *user_data;
+} oc_ns_t;
+
+
+/*
+ * structure for member of "oic.r.pushreceiver:receivers" object array
+ */
+typedef struct oc_recv
+{
+	struct oc_recv *next;
+	oc_string_t receiveruri;
+	oc_string_array_t rts;
+} oc_recv_t;
+
+
+/*
+ * Object creation and request handler for Push Receiver Resource
+ */
+typedef struct oc_recvs
+{
+	struct oc_recvs *next;
+	oc_resource_t *resource;
+	OC_LIST_STRUCT(receivers);
+} oc_recvs_t;
+
+
+
+/*
+ * object used to store Resource pushed to "oic.r.pshreceiver:receivers[i].receiveruri"
+ */
+typedef struct oc_pushd_rsc_rep
+{
+	struct oc_pushd_rsc_rep *next;
```

Comment:
add doxygen to the fields. 
for example 

/**
 * @brief device information
 *
 */
typedef struct oc_device_info_t
{
  oc_uuid_t di;                          ///< device indentifier
  oc_uuid_t piid;                        ///< Permanent Immutable ID
  oc_string_t name;                      ///< name of the device
  oc_string_t icv;                       ///< specification version
  oc_string_t dmv;                       ///< data model version
  oc_core_add_device_cb_t add_device_cb; ///< callback when device is changed
  void *data;                            ///< user data
} oc_device_info_t;